File: java-src/com/example/reviewdemo/order/OrderService.java
package com.example.reviewdemo.order;

public class OrderService {
    private final OrderRepository orderRepository;
    private final InventoryGateway inventoryGateway;
    private final PaymentGateway paymentGateway;

    public OrderService(
            OrderRepository orderRepository,
            InventoryGateway inventoryGateway,
            PaymentGateway paymentGateway
    ) {
        this.orderRepository = orderRepository;
        this.inventoryGateway = inventoryGateway;
        this.paymentGateway = paymentGateway;
    }

    public Order pay(String orderId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new IllegalArgumentException("order not found"));
        if (order.status() != OrderStatus.PENDING) {
            throw new IllegalStateException("order is not payable");
        }

        inventoryGateway.reserve(order.sku(), order.quantity());
        try {
            paymentGateway.charge(order.id(), order.amount());
            Order paid = order.markPaid();
            orderRepository.save(paid);
            return paid;
        } catch (RuntimeException error) {
            inventoryGateway.release(order.sku(), order.quantity());
            throw error;
        }
    }
}

File: src/order/order.service.ts
import type { PaymentGateway } from "../payment/payment.service.js";

export type Order = {
  id: string;
  sku: string;
  quantity: number;
  amountCents: number;
  status: "pending" | "paid";
};

export type InventoryStore = {
  reserve(sku: string, quantity: number): Promise<void>;
  release(sku: string, quantity: number): Promise<void>;
};

export class OrderService {
  constructor(
    private readonly gateway: PaymentGateway,
    private readonly inventory: InventoryStore,
  ) {}

  async pay(order: Order): Promise<Order> {
    await this.inventory.reserve(order.sku, order.quantity);
    try {
      await this.gateway.charge(order.id, order.amountCents);
      return {
        ...order,
        status: "paid",
      };
    } catch (error) {
      await this.inventory.release(order.sku, order.quantity);
      throw error;
    }
  }
}
